harness/determined/experimental/inference/_torch_batch_process.py

```diff
@@ -0,0 +1,412 @@
+        # Report reduced metrics to master
+        gatherables = [wrapped.per_slot_reduce() for wrapped in reducables]
+        if rank == 0:
+            gathered = core_context.distributed.gather(gatherables)
+            metrics = batch_processor_context.run_cross_slot_reduction(reducables, gathered)
+            core_context.train.report_validation_metrics(
+                steps_completed=steps_completed,
+                metrics=metrics,
+            )
+        else:
+            # Other ranks sent metrics to chief
+            core_context.distributed.gather(gatherables)
+
+
+def torch_batch_process(
+    batch_processor_cls: Type[TorchBatchProcessor],
+    dataset: Dataset,
+    batch_size: Optional[int] = None,
+    max_batches: Optional[int] = None,
+    checkpoint_interval: int = 5,
+    dataloader_kwargs: Dict[str, Any] = {},
+):
+    with initialize_default_inference_context() as core_context:
+        """
+        (1) Set up necessary variables to run batch processing
+        """
+
+        # Validate argument inputs
+        if checkpoint_interval <= 0:
+            raise ValueError("checkpoint_interval should be a positive integer")
+        _validate_dataloader_kwargs(dataloader_kwargs, batch_size)
+
+        if batch_size is None:
+            if "batch_size" in dataloader_kwargs:
+                # remove batch_size from dataloader_kwargs
+                # and assign to batch_size
+                batch_size = dataloader_kwargs.pop("batch_size")
+            else:
+                batch_size = DEFAULT_BATCH_SIZE
+        dataset_len = len(dataset)
+
+        info = det.get_cluster_info()
+
+        slots_per_node = len(info.slot_ids)
+        num_nodes = len(info.container_addrs)
+        total_worker = num_nodes * slots_per_node
+        # Get global rank
+        rank = core_context.distributed.get_rank()
+        latest_checkpoint = info.latest_checkpoint
+        skip = 0
+
+        # Synchronize default output uuid
+        if rank == 0:
+            default_output_uuid = str(uuid.uuid4())
+            core_context.distributed.broadcast(default_output_uuid)
+        else:
+            default_output_uuid = core_context.distributed.broadcast(None)
+
+        # Get previous trial state from checkpoint if available
+        if latest_checkpoint is not None:
+            logging.info("Checkpoint is not none")
+            with core_context.checkpoint.restore_path(latest_checkpoint) as path:
+                metadata = _load_state(path)
+                skip = metadata["steps_completed"]
+                logging.info(f"Previous run completed {skip} steps")
+                default_output_uuid = metadata["default_output_uuid"]
+
+        output_uuid_with_rank = default_output_uuid + f"/rank_{rank}"
+
+        batch_processor_context = TorchBatchProcessorContext(core_context, output_uuid_with_rank)
+
+        per_batch_processor = batch_processor_cls(
+            context=batch_processor_context,
+        )
```

Comment:
we've always gotten negative feedback about not allowing end users to instantiate their own class.

That said, Trainer API struggled a lot with coming up with a "nice" API where users could instantiate their own class, so I'm not sure what is best.

Reply:
> In our case, I am not sure why they would want to instantiate the processor class as it does not work outside of the torch_batch_process function

Right, I think that's kind of the point.  I've had lots of users tell me they are confused by the Trial because they can't "see what the entrypoint is", basically.

Admittedly, that feedback largely came at a time when the `entrypoint` in our system could only be the legacy `model_def:TrialClass` format, which definitely made things feel more opaque to end users.

Reply:
Eh, I guess even today, Garrett points out that a LightningModule is similar to a PyTorchTrial, only a LightningModule can be instantiated directly and you can make meaningful calls against it in a notebook.

Reply:
Chatted offline, this can be addressed when we receive feedback from user